rfcs/widget-restructure-rfc-1.md

```diff
@@ -0,0 +1,667 @@
+  styles: Styles,
+}
+
+// Generated:
+impl WidgetProps {
+  fn get_on_event(&self) -> Option<OnEvent> {
+  	Some(self.event_handler.clone())  
+  }
+  
+  fn get_styles(&self) -> Option<Style> {
+    None
+  }
+  
+  fn get_focusable(&self) -> Option<bool> {
+    self.focusable
+  }
+  
+  // ...
+}
+```
+
+#### The Constructor
+
+Another addition to `Widget` would be the constructor method. This associated method will allow widgets to be generated with a set of props in a more controlled and freeform way. The method looks like this:
+
+```rust
+fn constructor(id: Index, children: Children, mut props: Self::Props) -> dyn Widget<Props=Self::Props> where Self: Sized;
+```
+
+> Note that this would put the `Sized` restriction on widgets.
+
+Doing it this way allows the user to define the naming and placement of their props, or even replace them with their own.
+
+---
+
+### Defining Widgets
+
+#### Custom Widgets
+
+A custom widget is a widget defined manually by a user using standard Rust syntax (i.e. not by using the `#[widget]` attribute). By defining a widget manually, a user has finer control over behavior, data, and more.
+
+To define a custom widget, we need to first define our struct:
+
+```rust
+#[derive(Debug)]
+struct MyButton {
+  my_id: Index,
+  my_children: Children,
+  my_props: MyButtonProps,
+}
+```
+
+> We're naming these fields with the `my_` prefix to show that naming conventions are not mandatory for field names.
+
+##### Props
+
+While `Index` and `Children` are structs defined by Kayak, `MyWidgetProps` is not. This struct must be defined by the user and should contain all the props this widget expects.
+
+One vital bit about this struct is that it is the only part of this RFC that requires a **specific naming convention**. All props for a widget must be defined as `<Widget Name>Props`. So `Foo` contains `FooProps` and `Props` (an awful name for a widget) contains `PropsProps`.
+
+This is one of the major limitations to this system. Once the [`more_qualified_paths`](https://github.com/rust-lang/rust/issues/86935) feature is stabilized, we can remove this limitation by instead doing `<MyButton as Widget>::Props { ... }`. Unfortunately, this feature is only available in nightly at the moment.
+
+For now though we can create our props struct and derive `WidgetProps` as explained above:
+
+```rust
+#[derive(Debug, WidgetProps)]
+struct MyButtonProps {
+  // Kayak Props
+  #[props(OnEvent)]
+  event_handler: Option<OnEvent>,
+  #[props(Styles)]
+  styles: Option<Style>,
+  #[props(Focusable)]
+  focusable: bool,
+  
+  // Widget Props
+  text: String,
+  disabled: bool,
+}
+```
+
+##### Implementing Widget
+
+Now with everything defined, we're ready to implement `Widget`.
+
+Firstly, the constructor:
+
+```rust
+impl Widget for MyButton {
+  type Props = MyButtonProps;
+  
+  fn constructor(id: Index, children: Children, mut props: Self::Props) -> dyn Widget<Props=Self::Props> where Self: Sized {
```

Comment:
I'm not sure we actually have an id at this point. We normally get an ID when this is called:
`context.widget_manager.create_widget(0usize, built_widget, parent_id);`

Reply:
Gotcha, so should it be set to `Index::default()`?

Reply:
Yeah, that sounds right!